File: backend/src/main/java/com/app/starter1/persistence/entity/Order.java
package com.app.starter1.persistence.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "orders")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, name = "tenant_id")
    private Long tenantId;

    @Column(name = "customer_id")
    private Long customerId; // Opcional - puede ser venta sin cliente

    @Column(nullable = false, unique = true, length = 100, name = "invoice_number")
    private String invoiceNumber;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal subtotal;

    @Column(precision = 15, scale = 2)
    private BigDecimal tax;

    @Column(precision = 15, scale = 2)
    private BigDecimal discount;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal total;

    @Column(nullable = false, length = 50, name = "payment_method")
    private String paymentMethod; // CASH, CREDIT_CARD, DEBIT_CARD, TRANSFER

    @Column(nullable = false, length = 20)
    private String status; // COMPLETED, PENDING, CANCELLED

    @Column(name = "created_by")
    private Long createdBy; // Usuario que procesó la venta

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> items = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = this.createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Helper method para agregar items
    public void addItem(OrderItem item) {
        items.add(item);
        item.setOrder(this);
    }

    // Helper method para calcular total
    public void calculateTotal() {
        this.subtotal = items.stream()
                .map(OrderItem::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        if (this.discount == null) {
            this.discount = BigDecimal.ZERO;
        }
        if (this.tax == null) {
            this.tax = BigDecimal.ZERO;
        }

        this.total = this.subtotal.add(this.tax).subtract(this.discount);
    }
}

File: pos-python/models/order.py
class OrderItem:
    def __init__(self, product_id, product_name, quantity, price, discount, total):
        self.product_id = product_id
        self.product_name = product_name
        self.quantity = int(quantity)
        self.price = float(price)
        self.discount = float(discount)
        self.total = float(total)

    @classmethod
    def from_dict(cls, data):
        return cls(
            product_id=data.get('product_id'),
            product_name=data.get('product_name'),
            quantity=data.get('quantity'),
            price=data.get('price'),
            discount=data.get('discount', 0.0),
            total=data.get('total')
        )

class Order:
    def __init__(self, id, invoice_number, customer_id, customer_name, subtotal, discount, tax, total, payment_method, created_by, created_at, is_synced, items=None):
        self.id = id
        self.invoice_number = invoice_number
        self.customer_id = customer_id
        self.customer_name = customer_name
        self.subtotal = float(subtotal)
        self.discount = float(discount)
        self.tax = float(tax)
        self.total = float(total)
        self.payment_method = payment_method
        self.created_by = created_by
        self.created_at = created_at
        self.is_synced = int(is_synced)
        self.items = items if items is not None else []

    @classmethod
    def from_dict(cls, data):
        items_data = data.get('items', [])
        items = [OrderItem.from_dict(item) for item in items_data]
        return cls(
            id=data.get('id'),
            invoice_number=data.get('invoice_number'),
            customer_id=data.get('customer_id'),
            customer_name=data.get('customer_name', "Mostrador"),
            subtotal=data.get('subtotal'),
            discount=data.get('discount', 0.0),
            tax=data.get('tax', 0.0),
            total=data.get('total'),
            payment_method=data.get('payment_method'),
            created_by=data.get('created_by'),
            created_at=data.get('created_at'),
            is_synced=data.get('is_synced', 0),
            items=items
        )
